torchinductor/overrides.py

```diff
@@ -0,0 +1,162 @@
+import contextlib
+import logging
+import random
+import weakref
+
+import torch
+from torch import _prims
+
+log = logging.getLogger(__name__)
+
+
+@contextlib.contextmanager
+def patch_functions():
+    prior = torch.nn.functional.dropout
+    torch.nn.functional.dropout = lowmem_dropout
```

Comment:
@Chillee is there a better way to do decomposition *before* autograd?

For this I want to change the backwards formula to pass different things.  For now, I'm doing it at the python level, but I feel like this is something AOT Autograd should be able to do for me in a more robust way.

Reply:
If we had mechanism for swapping out functions before grad we could also auto-generate backwards from the forwards decomp.

Reply:
Hmm, not super easily. Maybe @ezyang has any thoughts?

In the past I considered tracing out the forwards pass first so we could transform the forwards pass first before transforming the backwards pass.

I also remember @ngimel had some concerns about doing this kind of dropout vs. just saving the seed directly.

I also wonder whether we should just do this kind of transformation in the graph. For example, if we were to rewrite the `randn_like` in the joint fw/bw graph to take a seed, and then allowed rematerializing `randn_like`, the partitioning algorithm would naturally work for this.

Reply:
Yeah with our typical approach (and the one nvfuser is using) making sure that with cuda graphs the rng will get into correct state is hard. In inductor we are setting the state via a separate call to `randint` (launches extra kernel) anyway, so it's entirely possible (but we are launching an extra kernel always, which the regular approach avoids)

Reply:
You don't have to monkeypatch the function, just define a TorchFunctionMode. But yes interposing at the Python level is the easiest way to get in before autograd. We have discussed more flexibility in interposing on dispatch but nothing implemented yet.

Reply:
@ezyang do you have an example of TorchFunctionMode usage?

Reply:
https://github.com/albanD/subclass_zoo/blob/main/autograd_monkeypatch.py and yes it will work with AOT Autograd